core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,233 @@
+   * out a barrier() call, and recycled on stage completed.
+   *
+   * @param barrierId Identifier of the barrier stage that make a barrier() call.
+   * @param numTasks Number of tasks of the barrier stage, all barrier() calls from the stage shall
+   *                 collect `numTasks` requests to succeed.
+   */
+  private class ContextBarrierState(
+      val barrierId: ContextBarrierId,
+      val numTasks: Int) {
+
+    // There may be multiple barrier() calls from a barrier stage attempt, `barrierEpoch` is used
+    // to identify each barrier() call. It shall get increased when a barrier() call succeed, or
+    // reset when a barrier() call fail due to timeout.
+    private var barrierEpoch: Int = 0
+
+    // An array of RPCCallContexts for barrier tasks that are waiting for reply of a barrier()
+    // call.
+    private val requesters: ArrayBuffer[RpcCallContext] = new ArrayBuffer[RpcCallContext](numTasks)
+
+    // A timer task that ensures we may timeout for a barrier() call.
+    private var timerTask: TimerTask = null
+
+    // Init a TimerTask for a barrier() call.
+    private def initTimerTask(): Unit = {
+      timerTask = new TimerTask {
+        override def run(): Unit = {
+          // Timeout current barrier() call, fail all the sync requests.
+          failAllRequesters(requesters, "The coordinator didn't get all barrier sync " +
+            s"requests for barrier epoch $barrierEpoch from $barrierId within ${timeout}s.")
+          cleanupBarrierStage(barrierId)
+        }
+      }
+    }
+
+    // Cancel the current active TimerTask and release resources.
+    private def cancelTimerTask(): Unit = {
+      if (timerTask != null) {
+        timerTask.cancel()
+        timerTask = null
+      }
+    }
+
+    // Process the global sync request. The barrier() call succeed if collected enough requests
+    // within a configured time, otherwise fail all the pending requests.
+    def handleRequest(requester: RpcCallContext, request: RequestToSync): Unit = synchronized {
+      val taskId = request.taskAttemptId
+      val epoch = request.barrierEpoch
+
+      // Require the number of tasks is correctly set from the BarrierTaskContext.
+      require(request.numTasks == numTasks, s"Number of tasks of $barrierId is " +
+        s"${request.numTasks} from Task $taskId, previously it was $numTasks.")
+
+      // Check whether the epoch from the barrier tasks matches current barrierEpoch.
+      logInfo(s"Current barrier epoch for $barrierId is $barrierEpoch.")
+      if (epoch != barrierEpoch) {
+        requester.sendFailure(new SparkException(s"The request to sync of $barrierId with " +
+          s"barrier epoch $barrierEpoch has already finished. Maybe task $taskId is not " +
+          "properly killed."))
+      } else {
+        // If this is the first sync message received for a barrier() call, start timer to ensure
+        // we may timeout for the sync.
```

Comment:
We create `ContextBarrierState` when we receive the first sync message, I think it's more clear to create the timer when creating `ContextBarrierState`, so that we don't need the `if` here.

Reply:
This is true if you have only one `barrier()` call in a barrier task. Otherwise, if the first `barrier()` call succeed, the RPCCallContext array shall get cleared and the barrier epoch increased, but we still reuse the same `ContextBarrierState` for forthcoming `barrier()` calls.

Reply:
ah i see, makes sense